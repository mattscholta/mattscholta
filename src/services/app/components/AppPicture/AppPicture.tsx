import * as React from 'react';

export type PictureAttributes = React.DetailedHTMLProps<
  React.HTMLAttributes<HTMLElement>,
  HTMLElement
>;

export interface AppPictureProps extends PictureAttributes {
  alt?: React.ImgHTMLAttributes<HTMLImageElement>['alt'];
  loading?: React.ImgHTMLAttributes<HTMLImageElement>['loading'];
  jpg?: React.SourceHTMLAttributes<HTMLElement>['srcSet'];
  png?: React.SourceHTMLAttributes<HTMLElement>['srcSet'];
  webp?: React.SourceHTMLAttributes<HTMLElement>['srcSet'];
}

/**
 * @name AppPicture
 * @external https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture
 * @description This component just wraps up the some of the messy markup
 * required for making use of modern web images... Only one image will be
 * shown the user, going from the top down, so we put the advanced "webp" up
 * top, then basic support of "srcSet" and lastly a fallback to a standard IMG
 */
export const AppPicture = (props: AppPictureProps) => {
  const { alt = '', className, jpg, loading, png, webp, ...rest } = props;

  if (!jpg && !png) {
    const message = 'A fallback image source of type JPG or PNG must be used';
    throw new Error(message);
  }

  return (
    <picture {...rest} className={className}>
      {webp && <source srcSet={webp} type="image/webp" />}
      {jpg && <source srcSet={jpg} type="image/jpeg" />}
      {png && !jpg && <source srcSet={png} type="image/png" />}
      <img alt={alt} loading={loading || 'lazy'} src={jpg || png} />
    </picture>
  );
};
