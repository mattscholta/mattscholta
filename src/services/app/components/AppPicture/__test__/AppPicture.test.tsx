import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';

import { AppPicture, AppPictureProps } from '../AppPicture';

describe('AppPicture', () => {
  let props: AppPictureProps;
  let wrapper: RenderResult;

  const createWrapper = () => <AppPicture {...props} />;

  const alt = 'Testing alt text, which is optional';
  const jpg = '/path/to/image.jpg';
  const png = '/path/to/image.png';

  beforeEach(() => {
    props = { alt, jpg };
    wrapper = render(createWrapper());
  });

  it(`supports jpg's`, () => {
    const img = wrapper.getByAltText(alt);

    expect(wrapper.container.childNodes).toMatchSnapshot('jpg');
    expect(img).toHaveAttribute('src', jpg);
  });

  it(`supports png's`, () => {
    props = { alt, png };
    wrapper.rerender(createWrapper());

    const img = wrapper.getByAltText(alt);

    expect(wrapper.container.childNodes).toMatchSnapshot('png');
    expect(img).toHaveAttribute('alt', alt);
    expect(img).toHaveAttribute('src', png);
  });

  it(`prefers jpg's over png's`, () => {
    props = { alt, jpg, png };
    wrapper.rerender(createWrapper());

    const img = wrapper.getByAltText(alt);

    expect(wrapper.container.childNodes).toMatchSnapshot('jpg vs. png');
    expect(img).toHaveAttribute('alt', alt);
    expect(img).toHaveAttribute('src', jpg);
  });

  it(`no alt text provided`, () => {
    props = { jpg };
    wrapper.rerender(createWrapper());

    const img = wrapper.baseElement.querySelector('img');

    expect(wrapper.container.childNodes).toMatchSnapshot('no alt text');
    expect(img).toHaveAttribute('alt', '');
    expect(img).toHaveAttribute('src', jpg);
  });

  // it('throws an error', () => {
  //   props = {}
  //
  //   try {
  //     wrapper.rerender(createWrapper())
  //   } catch (error) {
  //     expect(wrapper.container.childNodes).toMatchSnapshot();
  //   }
  // });
});
