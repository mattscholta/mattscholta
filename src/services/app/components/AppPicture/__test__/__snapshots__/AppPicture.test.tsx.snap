// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPicture no alt text provided: no alt text 1`] = `
NodeList [
  <picture>
    <source
      srcset="/path/to/image.jpg"
      type="image/jpeg"
    />
    <img
      alt=""
      loading="lazy"
      src="/path/to/image.jpg"
    />
  </picture>,
]
`;

exports[`AppPicture prefers jpg's over png's: jpg vs. png 1`] = `
NodeList [
  <picture>
    <source
      srcset="/path/to/image.jpg"
      type="image/jpeg"
    />
    <img
      alt="Testing alt text, which is optional"
      loading="lazy"
      src="/path/to/image.jpg"
    />
  </picture>,
]
`;

exports[`AppPicture supports jpg's: jpg 1`] = `
NodeList [
  <picture>
    <source
      srcset="/path/to/image.jpg"
      type="image/jpeg"
    />
    <img
      alt="Testing alt text, which is optional"
      loading="lazy"
      src="/path/to/image.jpg"
    />
  </picture>,
]
`;

exports[`AppPicture supports png's: png 1`] = `
NodeList [
  <picture>
    <source
      srcset="/path/to/image.png"
      type="image/png"
    />
    <img
      alt="Testing alt text, which is optional"
      loading="lazy"
      src="/path/to/image.png"
    />
  </picture>,
]
`;
